modules/javafx.graphics/src/main/native-glass/win/RoActivationSupport.cpp

```diff
@@ -0,0 +1,220 @@
+        loadFunction(hLibComBase, pWindowsCreateString, "WindowsCreateString") &&
+        loadFunction(hLibComBase, pWindowsDeleteString, "WindowsDeleteString");
+
+    if (!loaded) {
+        uninitializeRoActivationSupport();
+    } else {
+        HRESULT res = RoInitialize(RO_INIT_SINGLETHREADED);
+        if (FAILED(res)) {
+            fprintf(stderr, RoException("RoInitialize failed: ", res).message());
+            uninitializeRoActivationSupport();
+        } else {
+            initialized = true;
+        }
+    }
+}
+
+void uninitializeRoActivationSupport()
+{
+    if (initialized) {
+        RoUninitialize();
+    }
+
+    initialized = false;
+
+    if (hLibComBase) {
+        FreeLibrary(hLibComBase);
+        hLibComBase = NULL;
+        pRoInitialize = NULL;
+        pRoUninitialize = NULL;
+        pRoActivateInstance = NULL;
+        pWindowsCreateString = NULL;
+        pWindowsDeleteString = NULL;
+    }
+}
+
+bool isRoActivationSupported()
+{
+    return initialized;
+}
+
+HRESULT WINAPI RoInitialize(RO_INIT_TYPE initType)
+{
+    return pRoInitialize(initType);
+}
+
+void WINAPI RoUninitialize()
+{
+    pRoUninitialize();
+}
+
+HRESULT WINAPI RoActivateInstance(HSTRING activatableClassId, IInspectable** instance)
+{
+    return pRoActivateInstance(activatableClassId, instance);
+}
+
+HRESULT WINAPI WindowsCreateString(PCNZWCH sourceString, UINT32 length, HSTRING* string)
+{
+    return pWindowsCreateString(sourceString, length, string);
+}
+
+HRESULT WINAPI WindowsDeleteString(HSTRING string)
+{
+    return pWindowsDeleteString(string);
+}
+
+hstring::hstring(const char* str)
+{
+    int wstr_len = MultiByteToWideChar(CP_UTF8, 0, str, -1, nullptr, 0);
+    WCHAR* wstr = new WCHAR[wstr_len];
+    memset(wstr, 0, wstr_len * sizeof(WCHAR));
+    MultiByteToWideChar(CP_UTF8, 0, str, -1, wstr, wstr_len);
+    WindowsCreateString(wstr, wstr_len - 1, &hstr_);
+    delete[] wstr;
+}
+
+hstring::~hstring()
+{
+    WindowsDeleteString(hstr_);
+}
+
+hstring::operator HSTRING()
+{
+    return hstr_;
+}
+
+RoException::RoException(const char* message)
+{
+    if (message == nullptr) {
+        message = "";
+    }
+
+    // Copy the "message" string.
+    size_t len = strlen(message);
+    char* msg = new char[len + 1];
+    strcpy_s(msg, len + 1, message);
+    message_ = msg;
+}
+
+RoException::RoException(const char* message, HRESULT res)
+{
+    if (message == nullptr) {
+        message = "";
+    }
+
+    const wchar_t* error = _com_error(res).ErrorMessage();
+
+    // Concatenate the "message" and "error" strings.
+    int message_length = int(strlen(message));
+    int error_length = WideCharToMultiByte(CP_ACP, 0, error, -1, NULL, 0, NULL, FALSE);
+    char* result = new char[message_length + error_length];
+    WideCharToMultiByte(CP_ACP, 0, error, -1, result + message_length, error_length, NULL, FALSE);
+    memcpy_s(result, message_length, message, message_length);
```

Comment:
Same comment about error checking as in `hstring`.

Reply:
Done.